contracts/registry/src/lib.rs

```diff
@@ -723,12 +742,175 @@ mod tests {
+        assert!(!ctr.is_banned(alice2()));
+    }
+
+    #[test]
+    fn soul_transfer_no_tokens_from_caller() {
+        let (mut ctx, mut ctr) = setup(&issuer1(), 1 * MINT_DEPOSIT);
+        ctx.predecessor_account_id = alice();
+        testing_env!(ctx.clone());
+        assert!(!ctr.is_banned(alice()));
+        assert!(!ctr.is_banned(alice2()));
+        ctr.sbt_soul_transfer(alice2(), None);
+        assert!(ctr.is_banned(alice()));
+        assert!(!ctr.is_banned(alice2()));
+        // assert ban even is being emited after the caller with zero tokens has invoked the soul_transfer
+        let log_ban = mk_log_str("ban", &format!("[\"{}\"]", alice()));
+        assert_eq!(test_utils::get_logs(), log_ban);
+    }
+
+    #[test]
+    fn soul_transfer_limit() {
+        let max_gas: Gas = Gas::ONE_TERA.mul(300);
+        let (mut ctx, mut ctr) = setup(&issuer1(), 150 * MINT_DEPOSIT);
+        let batch_metadata = mk_batch_metadata(100);
+        assert!(batch_metadata.len() == 100);
 
-        // tests:
-        // + test soult transfer with "continuation" - use the internal _sbt_soul_transfer to control limit
-        // + edge case: caller doesn't have any token
-        // + edge case: soul_transfer resumed but there is no more tokens (must emit event in the last resumed call)
-        // + find all edge cases
+        // issuer_1
+        ctr.sbt_mint(vec![(alice(), batch_metadata[..50].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer1(), None), 50);
+
+        // issuer_2
+        ctx.predecessor_account_id = issuer2();
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(alice(), batch_metadata[50..].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice(), issuer2(), None), 50);
+
+        // add more tokens to issuer_1
+        ctx.predecessor_account_id = issuer1();
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(bob(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(bob(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(alice2(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(carol(), batch_metadata[..20].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(carol(), issuer1(), None), 20);
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(dan(), batch_metadata[..10].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(dan(), issuer1(), None), 10);
+
+        // soul transfer alice->alice2
+        ctx.predecessor_account_id = alice();
+
+        let limit: u32 = 25; //anything above this limit will fail due to exceeding maximum gas usage per call
```

Comment:
let's add a test to confirm that trying to transfer 30 tokens will fail with gas by using `should_panic`.
example:
```
    #[should_panic(expected = "less than or equal to 100")]
```
https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic

Reply:
DONE